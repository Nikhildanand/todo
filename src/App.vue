<template>
  <div
    data-cy="todo"
    class="todo"
  >
    <DoneBoard :done-items="items.doneItems" />
    <section
      class="todo__items"
      :class="{ todo_start: items.todoItems.length }"
    >
      <TodoList
        v-if="items.todoItems.length"
        :todo-items="items.todoItems"
        @check-todo-item="checkTodoItem"
      />
      <div
        v-else-if="loading"
        data-cy="loading"
      >
        Loading...
      </div>
      <div
        v-else-if="error"
        data-cy="error"
      >
        {{ error }}
      </div>
      <h2
        v-else
        data-cy="completed"
      >
        Hurray! I have cleared all todos.
      </h2>
    </section>
    <footer
      class="todo__footer"
      data-cy="footer"
    >
      Task Board by Nikhil D Anand
    </footer>
  </div>
</template>

<script setup>
import { onMounted, reactive, ref } from "vue";
import DoneBoard from "@/components/DoneBoard.vue";
import TodoList from "@/components/TodoList.vue";

const items = reactive({ todoItems: [], doneItems: [] });
const loading = ref(false);
const error = ref("");

onMounted(async () => {
  try {
    loading.value = true;
    const response = await fetch(`https://dummyjson.com/todos`);
    const data = await response.json();
    items.todoItems = data.todos;
  } catch (_) {
    error.value = "Oops! Unexpected error occurs!";
  } finally {
    loading.value = false;
  }
});

const checkTodoItem = (id) => {
  const todoItemIndex = items.todoItems.findIndex((item) => item.id === id);
  const newTodoItems = items.todoItems.splice(todoItemIndex, 1);
  items.doneItems = [...items.doneItems, ...newTodoItems];
};
</script>

<style scoped lang="scss">
.todo {
  display: grid;
  min-height: 100vh;
  padding: 1rem;
  text-align: center;

  &__items {
    align-self: center;
  }

  &__footer {
    align-self: end;
  }

  &_start {
    align-self: start;
  }
}
</style>
