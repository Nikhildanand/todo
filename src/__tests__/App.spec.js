import { render, screen } from "@testing-library/vue";
import { rest } from "msw";
import App from "../App.vue";

describe("App", () => {
  it("renders the footer text", async () => {
    const footerText = "Task Board by Nikhil D Anand";
    render(App);
    const app = await screen.findByText(footerText);
    expect(app.innerText).contains(footerText);
  });
  it("matches snapshot", async () => {
    const wrapper = render(App);
    expect(wrapper).toMatchSnapshot();
  });
  it("renders a loading message when the todo items are being loaded", async () => {
    const loadingText = "Loading...";
    render(App);
    const app = await screen.findByText(loadingText);
    expect(app.innerText).contains(loadingText);
  });

  it("renders an error message if there is an error while loading the todo items", async () => {
    render(App);

    rest.get("https://dummyjson.com/todos", (_, res) =>
      res.networkError("Failed to connect")
    );
    const expectedText = "Oops! Unexpected error occurs!";

    const value = await screen.findByText(expectedText);
    expect(value.innerText).toBe(expectedText);
  });

  it("renders a message if there are no todo items", async () => {
    render(App);

    rest.get("https://dummyjson.com/todos", (_, res, ctx) => res(ctx.json([])));
    const expectedText = "Hurray! I have cleared all todos.";
    const value = await screen.findByText(expectedText);
    expect(value.innerText).contains(expectedText);
  });
});
