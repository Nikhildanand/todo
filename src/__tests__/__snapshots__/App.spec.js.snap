// Vitest Snapshot v1

exports[`App > matches snapshot 1`] = `
{
  "baseElement": <body>
    <div>
      <div
        class="todo"
        data-cy="todo"
        data-v-7a7a37b1=""
      >
        <main
          data-v-7a7a37b1=""
          data-v-8afc1bf7=""
        >
          <h2
            data-v-8afc1bf7=""
          >
            Let's start
          </h2>
          <section
            class="board"
            data-v-8afc1bf7=""
          >
            <div
              class="board__grid"
              data-v-8afc1bf7=""
            >
              
              
            </div>
          </section>
        </main>
        <section
          class="todo__items"
          data-v-7a7a37b1=""
        >
          <h2
            data-cy="completed"
            data-v-7a7a37b1=""
          >
             Hurray! I have cleared all todos. 
          </h2>
        </section>
        <footer
          class="todo__footer"
          data-cy="footer"
          data-v-7a7a37b1=""
        >
           Task Board by Nikhil D Anand 
        </footer>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="todo"
      data-cy="todo"
      data-v-7a7a37b1=""
    >
      <main
        data-v-7a7a37b1=""
        data-v-8afc1bf7=""
      >
        <h2
          data-v-8afc1bf7=""
        >
          Let's start
        </h2>
        <section
          class="board"
          data-v-8afc1bf7=""
        >
          <div
            class="board__grid"
            data-v-8afc1bf7=""
          >
            
            
          </div>
        </section>
      </main>
      <section
        class="todo__items"
        data-v-7a7a37b1=""
      >
        <h2
          data-cy="completed"
          data-v-7a7a37b1=""
        >
           Hurray! I have cleared all todos. 
        </h2>
      </section>
      <footer
        class="todo__footer"
        data-cy="footer"
        data-v-7a7a37b1=""
      >
         Task Board by Nikhil D Anand 
      </footer>
    </div>
  </div>,
  "debug": [Function],
  "emitted": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "html": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
